### PR TITLE
[transformation] Anonymization

### PR DESCRIPTION
**Pull request by shananton**

### CODE REVIEW

File: docs/Anonymization.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+- `parentName` is the name of the entity's parent after renaming. 
+  The parent is usually the
+  most nested scope owner that the entity's definition resides in.
+  For example, the parent of a class method is the corresponding class.
+  Global entities do not have a parent and, for them, `parentName` is
+  considered to be empty.
+  
+- `kindPrefix` is one of the following, depending on what kind of entity
+  is being renamed:
+    | Prefix | Kind                    |
```
Comment (nbirillo) [inline docs/Anonymization.md:21]: Please, add an empty line before the table

File: docs/Anonymization.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    | ------ | ----                    |
+    | `f`    | Function                |
+    | `v`    | Variable                |
+    | `c`    | Class                   |
+    | `p`    | Parameter               |
+    | `m`    | Import alias ("module") |
+    | `l`    | Lambda expression       |
+  Note: lambda expressions do not have a name, thus cannot be renamed. However,
```
Comment (nbirillo) [inline docs/Anonymization.md:29]: Please, add an empty line after the table

File: docs/Anonymization.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  
+### Exceptions
+
+- Any entities whose name starts with `__` (two underscores)
+  are not renamed. This, in particular, preserves special method names
+  (`__init__` etc.) and [name mangling](https://docs.python.org/3/reference/expressions.html#atom-identifiers).
+- Any parameters of class methods that have one of the names `self` and `cls`
+  are not renamed.
+- Any entities that potentially redefine builtin variables and functions
+  (i. e. have the same name as a builtin and are global) are not renamed.
```
Comment (nbirillo) [inline docs/Anonymization.md:43]: I am not sure about this point. Maybe should we rename?

File: src/main/kotlin/org/jetbrains/research/ml/ast/transformations/anonymization/NamedEntityKind.kt [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+enum class NamedEntityKind(val prefix: String) {
+    Function("f"),
+    Variable("v"),
+    Class("c"),
+    Parameter("p"),
+    NamedImport("m"),
+    Lambda("l");
+    companion object {
+        fun computeElementKind(element: PsiElement): NamedEntityKind? = when (element) {
```
Comment (nbirillo) [inline src/main/kotlin/org/jetbrains/research/ml/ast/transformations/anonymization/NamedEntityKind.kt:18]: Maybe `getElementKind`?

**Review by nbirillo — CHANGES_REQUESTED** [review]
It looks totally great! Thank you!

File: docs/Anonymization.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+def foo(param):
+  inner = var
+```
+is transformed to:
+```
+v1 = 1
+v2 = 2
+
+def f1(f1p1):
+  f1v1 = v1
```
Comment (elena-lyulina) [inline docs/Anonymization.md:61]: Maybe we should have separators between the parent name and the other part to simplify reading? for example, f1_v1

File: src/main/kotlin/org/jetbrains/research/ml/ast/transformations/anonymization/AnonymizationVisitor.kt [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+    fun performAllRenames() {
+        val renames = anonymizer.getAllRenames().map { RenameUtil.renameElementDelayed(it.first, it.second) }
+        WriteCommandAction.runWriteCommandAction(project) {
+            renames.forEach { it() }
+        }
+    }
+
+    private val project = file.project
+    private val anonymizer = ElementAnonymizer()
```
Comment (elena-lyulina) [inline src/main/kotlin/org/jetbrains/research/ml/ast/transformations/anonymization/AnonymizationVisitor.kt:22]: Let's put the properties on top of the class according to documentation https://kotlinlang.org/docs/reference/coding-conventions.html#class-layout

**Review by elena-lyulina — CHANGES_REQUESTED** [review]
Nice

**Review by nbirillo — APPROVED** [review]
Great!

**Review by elena-lyulina — APPROVED** [review]

**Pull request merged**